src/compiler/crystal/semantic/exhaustiveness_checker.cr

```diff
@@ -0,0 +1,162 @@
+           when_cond.type?
+          cond_types.reject! { |type| type.implements?(when_cond.type.devirtualize) }
+          next
+        end
+
+        # At this point we are matching against other things than types/paths
+        all_whens_are_types = false
+
+        # If we are matching against an enum type and the Path denotes an enum member,
+        # delete it from the members we must cover
+        if when_cond.is_a?(Path) &&
+           enum_members && single_cond_type.is_a?(EnumType) &&
+           (target_const = when_cond.target_const)
+          matching_member = single_cond_type.types.find { |k, v| v == target_const }
+          if matching_member
+            enum_members.delete(matching_member[1].name.to_s)
+          end
+        end
+
+        case when_cond
+        when Call
+          # Check if it's something like `.foo?` to remove that member from the ones
+          # we must cover.
+          # Note: a user could override the meaning of such methods.
+          # In the future it would be wise to mark these as non-redefinable
+          # so this checks are sounds.
+          if enum_members && when_cond.obj.is_a?(ImplicitObj) &&
+             when_cond.args.empty? && when_cond.named_args.nil? &&
+             !when_cond.block && !when_cond.block_arg && when_cond.name.ends_with?('?')
+            enum_value = when_cond.name.rchop
+            enum_members.reject! { |value| value.underscore == enum_value }
+          end
+        when BoolLiteral
+          # Note that we found some bool literals
+          if when_cond.value
+            found_true = true
+          else
+            found_false = true
+          end
+        when NilLiteral
+          # A nil literal is the same as matching the Nil type
+          cond_types.reject! &.nil_type?
+        else
+          # Nothing to do
+        end
+      end
+    end
+
+    # If we found both `true` and `false` we covered Bool
+    if found_false && found_true
+      cond_types.delete(@program.bool)
+    end
+
+    # If we covered all types, we are done.
+    # This works even when matching against Bool or Enum.
+    return if cond_types.empty?
+
+    # If we didn't cover all types, but we tried to match against types,
+    # we know the user forgot some types.
+    if all_whens_are_types
+      @program.report_warning(node, <<-MSG)
+        case is not exhaustive.
+
+        Missing types: #{cond_types.join(", ")}
+        MSG
+      return
+    end
+
+    # Check the bool case
+    if single_cond_type.is_a?(BoolType) && !(found_false && found_true)
+      missing_cases = [] of String
+      missing_cases << "false" unless found_false
+      missing_cases << "true" unless found_true
+
+      @program.report_warning(node, <<-MSG)
+        case is not exhaustive.
+
+        Missing cases: #{missing_cases.join(", ")}
+        MSG
+      return
+    end
+
+    # Check the enum case
+    if single_cond_type.is_a?(EnumType) && enum_members
```

Comment:
Good catch. I can make it work, maybe even with several enums in the union type.

However, I think this is a pretty rare case. We should still support it and I can implement it, but I didn't want to do it before knowing whether we will continue this path or not.

Reply:
Again, seeking for clarification. I don't think it will be a primary use case. Yet, the unions of enums would look great :-P. No need for action here.